core/src/epoch_slots.rs

```diff
@@ -0,0 +1,396 @@
+use flate2::{Compress, Compression, Decompress, FlushCompress, FlushDecompress};
+use solana_sdk::clock::Slot;
+use solana_sdk::pubkey::Pubkey;
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
+pub struct Uncompressed {
+    pub first_slot: Slot,
+    pub num: usize,
+    pub slots: BitVec<u8>,
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq)]
+pub struct Flate2 {
+    pub first_slot: Slot,
+    pub num: usize,
+    pub compressed: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    CompressError,
+    DecompressError,
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl std::convert::From<flate2::CompressError> for Error {
+    fn from(_e: flate2::CompressError) -> Error {
+        Error::CompressError
+    }
+}
+impl std::convert::From<flate2::DecompressError> for Error {
+    fn from(_e: flate2::DecompressError) -> Error {
+        Error::DecompressError
+    }
+}
+
+impl Flate2 {
+    fn deflate(mut unc: Uncompressed) -> Result<Self> {
+        let mut compressed = Vec::with_capacity(unc.slots.block_capacity());
+        let mut compressor = Compress::new(Compression::best(), false);
+        let first_slot = unc.first_slot;
+        let num = unc.num;
+        unc.slots.shrink_to_fit();
+        let bits = unc.slots.into_boxed_slice();
+        compressor.compress_vec(&bits, &mut compressed, FlushCompress::Finish)?;
+        let rv = Self {
+            first_slot,
+            num,
+            compressed,
+        };
+        let _ = rv.inflate()?;
+        Ok(rv)
+    }
+    pub fn inflate(&self) -> Result<Uncompressed> {
+        //add some head room for the decompressor which might spill more bits
+        let mut uncompressed = Vec::with_capacity(32 + (self.num + 4) / 8);
+        let mut decompress = Decompress::new(false);
+        decompress.decompress_vec(&self.compressed, &mut uncompressed, FlushDecompress::Finish)?;
+        Ok(Uncompressed {
+            first_slot: self.first_slot,
+            num: self.num,
+            slots: BitVec::from_bits(&uncompressed),
+        })
+    }
+}
+
+impl Uncompressed {
+    pub fn new(max_size: usize) -> Self {
+        Self {
+            num: 0,
+            first_slot: 0,
+            slots: BitVec::new_fill(false, 8 * max_size as u64),
+        }
+    }
+    pub fn to_slots(&self, min_slot: Slot) -> Vec<Slot> {
+        let mut rv = vec![];
+        let start = if min_slot < self.first_slot {
+            0 as usize
+        } else {
+            (min_slot - self.first_slot) as usize
+        };
+        for i in start..self.num {
+            if self.slots.get(i as u64) {
+                rv.push(self.first_slot + i as Slot);
+            }
+        }
+        rv
+    }
+    pub fn add(&mut self, slots: &[Slot]) -> usize {
+        for (i, s) in slots.iter().enumerate() {
+            if self.num == 0 {
+                self.first_slot = *s;
+            }
+            if *s < self.first_slot {
+                return i;
```

Comment:
what if there are later slots in `slots` that fit in the range?

Reply:
It gets filled into the next instance of EpochSlots.  This one gets compressed.

Reply:
But won't this potentially cause unnecessary overlapping of ranges if we have a bad ordering of incoming slots? For example if the capacity is `10` and the ordering of slots is:

`1, 11 , 2, 12, 3, 13`

Then you've made 6 different slots in gossip when you could have fit them in 2

Reply:
@carllin they are compressed.  So worst case is that each slot has a unique entry.  But it’s unlikely because slots move forward and eventually first_slot will be the root of something.

Reply:
The range it starts with is something like 8000 bits.